### PR TITLE
[StackedBarPlotTests] - rework

### PR DESCRIPTION
**Pull request by bluong**

### CODE REVIEW

File: test/plots/stackedBarPlotTests.ts [hunk 1]
```diff
@@ -1,607 +1,411 @@
...
-        let actualY = TestMethods.numAttr(rect, "y");
-        assert.closeTo(actualY, expectedY, 0.1, "y attribute set correctly (dataset 1, datum " + index + ")");
-        let expectedHeight = yScale.scale(0) - expectedY;
-        let actualHeight = TestMethods.numAttr(rect, "height");
-        assert.closeTo(actualHeight, expectedHeight, 0.1, "height attribute set correctly (dataset 1, datum " + index + ")");
+      it("renders rects offset by previous values", () => {
+        let bars = stackedBarPlot.content().selectAll("rect");
+
+        let dataLength = stackedBarPlot.datasets()[0].data().length;
+        let dataCount = stackedBarPlot.datasets().length * dataLength;
```
Comment (ztsai) [inline test/plots/stackedBarPlotTests.ts]: this assumes datasets all have the same length

File: test/plots/stackedBarPlotTests.ts [hunk 2]
```diff
@@ -1,607 +1,411 @@
...
+      it("doesn't show any off-bar labels", () => {
+        stackedBarPlot.labelsEnabled(true);
+        yScale.domain([0, 30]);
+        let offBarLabels = stackedBarPlot.content().selectAll(".off-bar-label");
+        assert.operator(offBarLabels.size(), ">", 0, "some off-bar labels are drawn");
+        let offBarLabelVisibilities: string[] = [];
+        offBarLabels.each(function() {
+          offBarLabelVisibilities.push(d3.select(this).style("visibility"));
+        });
+        assert.isTrue(offBarLabelVisibilities.every((visibility) => visibility === "hidden"), "all off-bar labels are hidden");
```
Comment (ztsai) [inline test/plots/stackedBarPlotTests.ts]: is there any reason why you didn't test `visibility === "hidden"` in `each` above but push them to an array then test `every`?

File: test/plots/stackedBarPlotTests.ts [hunk 3]
```diff
@@ -1,607 +1,411 @@
...
+          y: yScale.scale(d0.y)
+        };
+
+        let d1 = stackedBarPlot.datasets()[1].data()[0];
+        let d1Px = {
+          x: xScale.scale(d1.x),
+          y: yScale.scale(d0.y + d1.y)
+        };
+
+        let expected: Plottable.Plots.PlotEntity = {
```
Comment (ztsai) [inline test/plots/stackedBarPlotTests.ts]: is it necessary to create a whole entity? is checking datum/index of the selected entity not sufficient?

File: test/plots/stackedBarPlotTests.ts [hunk 4]
```diff
@@ -1,607 +1,411 @@
...
-      assert.closeTo(TestMethods.numAttr(cBars[0], "x"), TestMethods.numAttr(cBars[1], "x"), 0.01, "C bars at same x position");
-      assert.operator(TestMethods.numAttr(cBars[0], "y"), ">", TestMethods.numAttr(cBars[1], "y"), "first dataset C bar under second");
-
-      svg.remove();
+    describe("fail safe tests", () => {
+
+      it("should default to 0 when calculating stack offsets with non-numbers", () => {
+        let svg = TestMethods.generateSVG();
+        let data1 = [
+          { x: "A", y: "s"},
```
Comment (ztsai) [inline test/plots/stackedBarPlotTests.ts]: what about null, undefined, Infinity, -Infinity, NaN?

File: test/plots/stackedBarPlotTests.ts (hunk 1 above)
Comment (bluong) [inline test/plots/stackedBarPlotTests.ts]: there is a separate `describe` block for non-overlapping datasets

File: test/plots/stackedBarPlotTests.ts (hunk 2 above)
Comment (bluong) [inline test/plots/stackedBarPlotTests.ts]: This was how the old test is written.  I do agree that it is unconventional and possibly harder to read.  I'll use `.each` instead

File: test/plots/stackedBarPlotTests.ts (hunk 3 above)
Comment (bluong) [inline test/plots/stackedBarPlotTests.ts]: This was how the old test was written.  I'll look into if this test ca be more clean

File: test/plots/stackedBarPlotTests.ts (hunk 4 above)
Comment (bluong) [inline test/plots/stackedBarPlotTests.ts]: `Infinity` and `-Infinity` are both `number`s.  I can test the other 3

Comment (ztsai) [inline test/plots/stackedBarPlotTests.ts]: well, technically, `NaN` is `number` as well
just saying those are edge cases to be tested against

File: test/plots/stackedBarPlotTests.ts [hunk 5]
```diff
@@ -1,607 +1,411 @@
...
-      let cBars = [d3.select(bars[0][2]), d3.select(bars[0][6])];
-
-      assert.closeTo(TestMethods.numAttr(aBars[0], "y"), TestMethods.numAttr(aBars[1], "y"), 0.01, "A bars at same y position");
-      assert.operator(TestMethods.numAttr(aBars[0], "x"), "<", TestMethods.numAttr(aBars[1], "x"), "first dataset A bar under second");
-
-      assert.closeTo(TestMethods.numAttr(bBars[0], "y"), TestMethods.numAttr(bBars[1], "y"), 0.01, "B bars at same y position");
-      assert.closeTo(TestMethods.numAttr(bBars[1], "y"), TestMethods.numAttr(bBars[2], "y"), 0.01, "B bars at same y position");
-      assert.operator(TestMethods.numAttr(bBars[0], "x"), "<", TestMethods.numAttr(bBars[1], "x"), "first dataset B bar under second");
-      assert.operator(TestMethods.numAttr(bBars[1], "x"), "<", TestMethods.numAttr(bBars[2], "x"), "second dataset B bar under third");
+  describe("Horizontal Stacked Bar Plot", () => {
```
Comment (ztsai) [inline test/plots/stackedBarPlotTests.ts]: fail safe tests for horizontal stacked bars?

File: test/plots/stackedBarPlotTests.ts [hunk 6]
```diff
@@ -1,607 +1,440 @@
...
-        "fifth bar should have offset 3, because fourth bar was not rendered");
+    describe("fail safe tests", () => {
+
+      it("should default to 0 when calculating stack offsets with non-numbers", () => {
+        let svg = TestMethods.generateSVG();
+        let stringData = [
+          { x: "s", y: "A"},
+        ];
+        let nullData = [
+          { x: <any> null, y: "A"},
```
Comment (ztsai) [inline test/plots/stackedBarPlotTests.ts]: why is any case needed?
